graphqlapi/src/main/java/io/stargate/graphql/schema/FieldTypeCache.java

```diff
@@ -0,0 +1,95 @@
+package io.stargate.graphql.schema;
+
+import graphql.Scalars;
+import graphql.schema.GraphQLScalarType;
+import io.stargate.db.schema.Column.ColumnType;
+import io.stargate.db.schema.Column.Type;
+import io.stargate.db.schema.ImmutableListType;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Caches a category of GraphQL field types, corresponding to table columns or UDT fields.
+ *
+ * <p>There are different categories, each implemented by a subclass. Note that this cache does not
+ * contain table types, they are handled as top-level entities in {@link DmlSchemaBuilder}.
+ *
+ * @param <GraphqlT> the returned GraphQL type.
+ */
+abstract class FieldTypeCache<GraphqlT> {
+
+  protected final NameMapping nameMapping;
+  private final Map<ColumnType, GraphqlT> types = new HashMap<>();
+
+  FieldTypeCache(NameMapping nameMapping) {
+    this.nameMapping = nameMapping;
+  }
+
+  GraphqlT get(ColumnType type) {
+    type = normalize(type);
+    return types.computeIfAbsent(type, this::compute);
+  }
+
+  private ColumnType normalize(ColumnType type) {
+    // Frozen-ness does not matter. We want frozen and non-frozen versions of a CQL type to be
+    // mapped to the same GraphQL type.
+    type = type.frozen(false);
```

Comment:
`frozen()` does not modify the type in place, it creates a copy. So this will only affect the cache key.

Currently we generate the same GraphQL type whether the CQL type is frozen or not, so we have to do this otherwise we have duplicate types and GraphQL will crash (try to comment this line, you'll see that some unit tests break).

If frozen-ness starts to leak in the GraphQL schema, we'll have to modify the caches anyway, so we'll consider what information to add in the key at that moment.

Reply:
oh snap! I haven't checked that `frozen()` method returned a new type instance.
Thanks for the explanation, nvm then.